File: 2015-ruby/lib/day_01.rb
require "set"

class Day01
  def part1(input)
    input.split(//).sum do |v| 
      case v
      when "(" then 1
      when ")" then -1
      end
    end
  end

  def part2(input)
    input.split(//).each_with_index.reduce(0) do |acc, (v, i)|
      acc += case v
      when "(" then 1
      when ")" then -1
      end

      break (i + 1) if acc < 0

      acc
    end
  end
end

File: 2022/lib/day_01.ex
defmodule Day01 do
  def part1(input) do
    input
    |> elf_calories()
    |> Enum.max()
  end

  def part2(input) do
    input
    |> elf_calories()
    |> Enum.sort(:desc)
    |> Enum.take(3)
    |> Enum.sum()
  end

  defp elf_calories(input) do
    input
    |> String.split("\n\n")
    |> Enum.map(fn elf ->
      elf
      |> String.split()
      |> Enum.map(&String.to_integer/1)
      |> Enum.sum()
    end)
  end
end
